stack/__main__.py

```diff
@@ -331,13 +331,76 @@ def create_group(mail: str) -> gcp.cloudidentity.Group:
         member=f'serviceAccount:{WEB_SERVER_SERVICE_ACCOUNT}',
     )
 
-    for group_prefix in ACCESS_LEVELS:
-        gcp.secretmanager.SecretIamMember(
-            f'sample-metadata-api-{group_prefix}-access-group-cache-secret-accessor',
-            secret_id=access_group_cache_secrets[group_prefix].id,
-            role='roles/secretmanager.secretAccessor',
-            member=f'serviceAccount:{SAMPLE_METADATA_API_SERVICE_ACCOUNT}',
-        )
+    # Sample metadata access
+    #   - 4 secrets, production-read, production-write, development-read, development-write
+    sm_groups = {}
+    for env in ('production', 'test'):
+        for rs in ('read', 'write'):
+            key = f'sample-metadata-{env}-{rs}'
+            secret_id = f'{dataset}-{key}-members-cache'
+
+            group = create_group(group_mail(dataset, key))
+            sm_groups[f'{env}-{rs}'] = group
+
+            gcp.cloudidentity.GroupMembership(
+                f'sample-metadata-group-cache-{env}-{rs}-access-level-group-membership',
+                group=group,
+                preferred_member_key=gcp.cloudidentity.GroupMembershipPreferredMemberKeyArgs(
+                    id=SAMPLE_METADATA_API_SERVICE_ACCOUNT
+                ),
+                roles=[gcp.cloudidentity.GroupMembershipRoleArgs(name='MEMBER')],
+                opts=pulumi.resource.ResourceOptions(depends_on=[cloudidentity]),
+            )
+
+            gcp.secretmanager.Secret(
+                f'{key}-group-cache-secret',
+                secret_id=secret_id,
+                replication=gcp.secretmanager.SecretReplicationArgs(
+                    user_managed=gcp.secretmanager.SecretReplicationUserManagedArgs(
+                        replicas=[
+                            gcp.secretmanager.SecretReplicationUserManagedReplicaArgs(
+                                location='australia-southeast1',
+                            ),
+                        ],
+                    ),
+                ),
+                opts=pulumi.resource.ResourceOptions(depends_on=[secretmanager]),
+            )
+
+            gcp.secretmanager.SecretIamMember(
+                f'sample-metadata-api-{env}-{rs}-access-group-cache-secret-accessor',
+                secret_id=secret_id,
+                role='roles/secretmanager.secretAccessor',
+                member=f'serviceAccount:{SAMPLE_METADATA_API_SERVICE_ACCOUNT}',
+            )
+
+    sm_access_levels = [
```

Comment:
I think this structure warrants a comment. Or maybe you could use a `NamedTuple` to make this clearer?

Reply:
Yep, good idea. I've added a `namedtuple` and used `kwargs`, it's a bit more verbose but more clear.